'''
Created on Jan 16, 2026

Convert transcript effects json to csv 

@author: pleyte
'''
import logging.config
import argparse
from rinc.util.log_config import LogConfig
from rinc.variant_transcript import VariantTranscript
import json
from rinc.util import chromosome_map, vcf_to_gdot
from collections import Counter
from rinc.io import variant_helper
from rinc.util.tx_eff_pysam import PysamTxEff

class TfxNomenclature(object):
    '''
    classdocs
    '''
    def __init__(self, fasta_file):
        '''
        Constructor
        '''
        self._logger = logging.getLogger(__name__)
        self._pysam_tx = PysamTxEff(fasta_file)

    def _get_variant_transcript(self, j:dict):
        """
        Create VariantTrancript using a dict of Tfx values
        """
        v = VariantTranscript(j['chromosome'], j['position'], j['reference'], j['alt'], j['cdnaTranscript'])
        
        v.c_dot = j['cdnaNomenclature']
        v.exon = j['exon']
        
        if j['sequenceVariant']:            
            g_dot_part = j['sequenceVariant']        
            refseq_chromosome = chromosome_map.get_refseq(v.chromosome)
            v.g_dot = f"{refseq_chromosome}:{g_dot_part}"
        
        v.gene = j['gene']
        v.genomic_region_type = j['variantType']
        v.p_dot1 = j['proteinNomenclature1']
        v.p_dot3 = j['proteinNomenclature3']
        v.protein_transcript = j['proteinTranscript']
        v.protein_variant_type = j['variantEffect']
        v.additional_fields['splicing'] = j['splicing']
        v.additional_fields['vrnt_type'] = self._get_variant_type(v)
        
        return v
        
    def _get_variant_type(self, v: VariantTranscript) -> str:
        """
        When transcript effects was generating its g_dot, using vcf_to_gdot.py, it determined the variant type.
        We want to know what that variant type is, so we're calling vcf_to_gdot.get_gdot_plus to get it.         
        """
        g_dot, vrnt_type = vcf_to_gdot.get_gdot_plus(v.chromosome, v.position, v.reference, v.alt, self._pysam_tx)
        if v.g_dot and g_dot != v.g_dot:
            raise ValueError(f"G-dot do not match even though they were generated by the same function: {g_dot} != {v.g_dot}")
        
        return vrnt_type
        
    def get_variant_transcripts(self, tfx_json_file: str) -> list[VariantTranscript]:
        """
        """
        count = Counter()
        with open(tfx_json_file, 'r') as f:
            data = json.load(f)
        
        self._logger.info(f"Read {len(data)} tfx from {tfx_json_file}")
        
        variant_transcripts = []
        
        for x in data['transcriptEffects']:
            if not x['cdnaTranscript']:
                count['skip_no_transcript'] += 1
                self._logger.info(f"Skipping variant that doesn't have transcripts: {x['chromosome']}-{x['position']}-{x['reference']}-{x['alt']}-{x['cdnaTranscript']}")
                continue
                        
            v = self._get_variant_transcript(x)            
            variant_transcripts.append(v)
            count['add_refseq_transcript'] += 1
        
        self._logger.info(f"{count}")
        return variant_transcripts
    
    def write(self, out_filename, variant_transcripts: list[VariantTranscript]):
        """
        """
        variant_helper.write_variant_transcripts(out_filename, variant_transcripts, ['splicing', 'vrnt_type'])
        self._logger.info(f"Wrote {len(variant_transcripts)} variant transcripts to {out_filename}")

def _parse_args():
    parser = argparse.ArgumentParser(description='Use hgvs to determine protein changes for a list of variants')
    parser.add_argument("--version", action="version", version="0.0.1")
    parser.add_argument("--fasta", help="hg fasta", required=True)
    parser.add_argument("--tfx_input", help="File generated by Transcript Effects (json)", required=True)
    parser.add_argument("--out", help="Tfx nomenclature output file (csv)", dest="output", required=True)
    args = parser.parse_args()
    return args    

def main():
    logging.config.dictConfig(LogConfig().stdout_config)

    args = _parse_args()

    tn = TfxNomenclature(args.fasta)
    
    variant_transcripts = tn.get_variant_transcripts(args.tfx_input)
    tn.write(args.output, variant_transcripts)
    
if __name__ == '__main__':
    main()                
