'''
Extract transcript details from an annovar multianno file and write to a new csv. 

Created on Jan 5, 2026

@author: pleyte
'''
import argparse
import csv
import logging.config
from rinc.util.log_config import LogConfig
from rinc.variant_transcript import VariantTranscript
import re

class AnnovarMultianno(object):
    '''
    classdocs
    '''
    def __init__(self):
        '''
        Constructor
        '''
        self._logger = logging.getLogger(__name__)
    
    def get_variant_transcripts(self, annovar_multianno_file):
        """
        Read the Annovar multianno file generated by ``table_annovar.pl``.         
        """
        variant_transcripts = []  
        with open(annovar_multianno_file, mode='r', encoding='utf-8') as f:        
            reader = csv.DictReader(f)
            for row in reader:
                for row_transcript in self._get_row_transcripts(row['Gene.refGeneWithVer']):
                    variant_transcript = self._get_parse_transcript(row_transcript, row)
                    variant_transcripts.append(variant_transcript)
        
        return variant_transcripts
    
    def _get_row_transcripts(self, gene_ref_gene_with_ver):
        """
        The annovar multianno file's Gene.refGeneWithVer lists the transcripts at this location. 
        The transcripts may be associated with multiple genes so there may two delimiters used as in: 
        NM_001242366.3,NM_080669.6;NM_015077.4
        
        The semi-colon delimits transcripts associated with different genes. 
        The comma delimts transcripts on the same gene.  
        """
        all_transcripts = []
        
        gene_groups = gene_ref_gene_with_ver.split(';')
        for g in gene_groups:
            all_transcripts.extend(g.split(','))
        
        return all_transcripts
        
        
    def _get_cdna_values(self, transcript: str, gene_detail_field: str):
        """
        The Annovar GeneDetail.refGeneWithVer field which is delimited by ';' and ':'
        """
        if gene_detail_field == '.':
            return None, None
                
        try:
            for gene_detail in gene_detail_field.split(';'):
                for gene_transcript_detail in gene_detail.split(","):
                    if 'exon' in gene_transcript_detail:            
                        gene_transcript, exon, c_dot = gene_transcript_detail.split(':')
                    elif gene_transcript_detail.startswith('dist='):
                        # dist=nnn means intergenic. We skipt these in tfx. 
                        return None, None
                    else:
                        exon = None
                        gene_transcript, c_dot = gene_transcript_detail.split(':')
                        
                    if gene_transcript == transcript:
                        return c_dot, exon 
        except ValueError:
            self._logger.warning(f"Could not parse GeneDetail.refGeneWithVer field for {transcript}: {gene_detail_field}")            
            raise
                
        # No information in this field for the current transcript 
        return None, None
        
    def _get_protein_values(self, transcript: str, aa_change_field: str):
        """
        Parse details for a specific transcript from the Annovar AAChange.refGeneWithVer field which is delimited by ',' and ':' and sometimes even a third delimiter ';'
        eg with pdot: "MAGT1:NM_001367916.1:exon2:c.C127T:p.Q43X,MAGT1:NM_032121.5:exon2:c.C223T:p.Q75X"
        or w/o pdot: "TNFRSF14:NM_003820.3:exon1:c.-13_1delinsC,TNFRSF14:NM_001297605.1:exon1:c.-13_1delinsC"
        
        """
        for aa_detail in re.split(r'[,;]', aa_change_field): 
            if aa_detail == "." or aa_detail == "UNKNOWN":
                continue
            
            gene, aa_transcript, exon, c_dot, *rest = aa_detail.split(':')
            
            if rest and len(rest) > 1:
                raise ValueError(f"annovar string has more than five parts: {aa_change_field}")
            elif rest:
                p_dot = rest[0]
            else:
                p_dot = None

            if aa_transcript == transcript:
                return gene, exon, c_dot, p_dot
        
        # No information in this field for the current transcript 
        return None, None, None, None
            
    def _normalize(self, c_dot_raw) -> str:
        """
        Attempt to normalize annovar's c.
        Supported formats are:
        * Substitutions: c.C1307T --> c.1307C>T
        
        """        
        substitution_pattern = r"c\.(?P<ref>[A-Z])(?P<pos>\d+)(?P<alt>[A-Z])"
        match = re.search(substitution_pattern, c_dot_raw)
        if match:
            ref_base = match.group("ref")
            position = int(match.group("pos"))
            alt_base = match.group("alt")
            return f"c.{position}{ref_base}>{alt_base}"
        
        return c_dot_raw
        
    def _get_parse_transcript(self, transcript: str, row: dict):
        """
        Extract details for a specific transcript from an annovar row
        """
        variant_transcript = VariantTranscript(row['Chr'], row['Start'], row['Ref'], row['Alt'], transcript)
        variant_transcript.genomic_region_type = row['Func.refGeneWithVer']
        variant_transcript.protein_variant_type = row['ExonicFunc.refGeneWithVer']
                    
        cdna_c_dot, cdna_exon = self._get_cdna_values(transcript, row['GeneDetail.refGeneWithVer'])
        aa_gene, aa_exon, aa_c_dot, aa_p_dot = self._get_protein_values(transcript, row['AAChange.refGeneWithVer'])
        
        # Get the c. from one of two fields
        if cdna_c_dot and aa_c_dot:
            variant_transcript.c_dot = cdna_c_dot
            variant_transcript.notes.append(f"c. from protein field is {aa_c_dot}")
        elif cdna_c_dot:
            variant_transcript.c_dot = cdna_c_dot
        elif aa_c_dot:
            variant_transcript.additional_fields['aa_c_dot_raw'] = aa_c_dot 
            variant_transcript.c_dot = self._normalize(aa_c_dot)
            variant_transcript.notes.append(f"c. is from protein field. See aa_c_dot_raw")
        
        # Exon
        if cdna_exon and aa_exon:
            variant_transcript.exon = cdna_exon
            variant_transcript.notes.append(f"exon from protein field is {aa_exon}")
        elif cdna_exon:
            variant_transcript.exon = cdna_exon
        elif aa_exon:
            variant_transcript.exon = aa_exon
        
        
        # Gene
        variant_transcript.gene = aa_gene
        
        # p.
        variant_transcript.p_dot1 = aa_p_dot

        # 
        variant_transcript.notes.append("source=annovar")
        
        return variant_transcript

    def _get_exon(self, exon: str):
        """
        Convert annovar's exon string like "exon1" to just "1"
        """
        if exon:
            return exon.replace('exon', '')
        else:
            return exon
        
        
    def write(self, out_filename, variant_transcripts: list[VariantTranscript]):
        headers = ['chromosome', 'position', 'reference', 'alt',
                   'cdna_transcript',
                   'annovar.exon', 'annovar.gene', 'annovar.genomic_region_type', 'annovar.protein_variant_type',
                   'annovar.c_dot', 'annovar.p_dot1',
                   'annovar.note']
        
        with open(out_filename, 'w', newline='') as output:
            writer = csv.writer(output)
            writer.writerow(headers)

            for v in variant_transcripts:
                writer.writerow([v.chromosome, v.position, v.reference, v.alt, 
                                 v.cdna_transcript, 
                                 self._get_exon(v.exon), v.gene, v.genomic_region_type, v.protein_variant_type,
                                 v.c_dot, v.p_dot1, 
                                 ";".join(v.notes)])
            
        self._logger.info(f"Wrote {len(variant_transcripts)} variant transcripts to {out_filename}")


def _parse_args():
    parser = argparse.ArgumentParser(description='Read Annovar multianno file, extract values and write to new csv')

    parser.add_argument('--annovar_multianno',
                        help='Multianno output from annovar (csv)',
                        required=True)

    parser.add_argument("--out", help="output file (csv)", required=True)

    parser.add_argument("--version", action="version", version="0.0.1")

    return parser.parse_args()


def main():
    logging.config.dictConfig(LogConfig().stdout_config)
    logger = logging.getLogger("rinc.etl.annovar_multianno")

    args = _parse_args()

    am = AnnovarMultianno()
    
    variant_transcripts = am.get_variant_transcripts(args.annovar_multianno)

    am.write(args.out, variant_transcripts)

    logger.info(f"Wrote {len(variant_transcripts)} variant transcripts to {args.out}")


if __name__ == '__main__':
    main()
