'''
Read a tsv generated by SnpEff, and write out a csv 
Created on Jan 10, 2026

@author: pleyte
'''
import argparse
import csv
import logging.config
from rinc.util.log_config import LogConfig
from rinc.variant_transcript import VariantTranscript
from rinc.util.pdot import PDot

class ProcessSnpeff(object):
    '''
    classdocs
    '''

    def __init__(self):
        '''
        Constructor
        '''
        self._logger = logging.getLogger(__name__)
        self._pd = PDot()
    
    def get_variant_transcripts(self, snpeff_tsv_file):
        """
        Read the SnpEff tsv file generated by snpEff.jar, vcfEffOnePerLine.pl, and /SnpSift.jar.         
        """
        variant_transcripts = {} 
        
        with open(snpeff_tsv_file, mode='r', encoding='utf-8') as f:        
            reader = csv.DictReader(f, delimiter='\t')
            for row in reader:
                if not row['ANN[*].FEATUREID'].startswith('NM_'):
                    #self._logger.debug(f"Skipping transcript {row['ANN[*].FEATUREID']}")
                    continue
                
                vt = VariantTranscript(row['CHROM'], row['POS'], row['REF'], row['ALT'], row['ANN[*].FEATUREID'])
                
                vt.gene = row['ANN[*].GENE']
                vt.exon = row['ANN[*].RANK']
                vt.c_dot = row['ANN[*].HGVS_C']
                vt.p_dot3 = row['ANN[*].HGVS_P']
                
                if vt.p_dot3 and vt.p_dot3 != '.':
                    vt.p_dot1 = self._pd.get_p_dot1(vt.cdna_transcript, vt.p_dot3)
                    
                vt.additional_fields['ann_effect_raw'] = row['ANN[*].EFFECT']
                vt.additional_fields['ann_biotype_raw'] = row['ANN[*].BIOTYPE']
                
                vt.genomic_region_type = self._get_genomic_region_type(row['ANN[*].EFFECT']) 
                
                key = f"{vt.chromosome}-{vt.position}-{vt.reference}-{vt.alt}-{vt.cdna_transcript}"
                vt_existing = variant_transcripts.get(key) 
                if vt_existing:
                    self._logger.info(f"Duplicate row for {key}, skipping the latter: {vt_existing.c_dot} ? {vt.c_dot}, {vt_existing.p_dot1} ? {vt.p_dot1}, {vt_existing.exon} ? {vt.exon}")
                else:
                    variant_transcripts[key] = vt
        
        return variant_transcripts.values()
    
    def _get_genomic_region_type(self, snpeff_effect: str):
        """
        Convert SnpEff EFFECT to "exon", "intron", etc
        Full list of EFFECT values: https://pcingola.github.io/SnpEff/snpeff/inputoutput/#eff-field-vcf-output-files        
        """
        effect = snpeff_effect.lower()
        if effect in [ "missense_variant", "synonymous_variant", "stop_gained", "stop_lost", 
                             "start_lost", "frameshift_variant", "inframe_insertion", "inframe_deletion",
                             "non_coding_transcript_exon_variant", "non_coding_exon_variant", 
                             "missense_variant&splice_region_variant", "splice_region_variant&synonymous_variant",
                             "non_coding_transcript_exon_variant" ]:
            return "exon"
        
        if "intron_variant" in effect:
            return "intron"
        
        if "utr_variant" in effect or effect == "5_prime_UTR_premature_start_codon_gain_variant":
            return "utr"
                
        if effect in ['intergenic_region', 'downstream_gene_variant', 'upstream_gene_variant']:
            return "intergenic"
        
        self._logger.info(f"Unknown SnpEff EFFECT: {snpeff_effect}")
        return None
    
    def write(self, out_filename, variant_transcripts: list[VariantTranscript]):
        headers = ['chromosome', 'position', 'reference', 'alt',
                   'cdna_transcript',
                   'snpeff.exon', 'snpeff.gene', 'snpeff.genomic_region_type',
                   'snpeff.c_dot', 'snpeff.p_dot1', 'snpeff.p_dot3',
                   'ann_effect_raw', 'ann_biotype_raw']
        
        with open(out_filename, 'w', newline='') as output:
            writer = csv.writer(output)
            writer.writerow(headers)

            for v in variant_transcripts:
                writer.writerow([v.chromosome, v.position, v.reference, v.alt, 
                                 v.cdna_transcript, 
                                 v.exon, v.gene, v.genomic_region_type, 
                                 v.c_dot, v.p_dot1, v.p_dot3, 
                                 v.additional_fields['ann_effect_raw'], 
                                 v.additional_fields['ann_biotype_raw']])
            
        self._logger.info(f"Wrote {len(variant_transcripts)} variant transcripts to {out_filename}")
        

        
def _parse_args():
    parser = argparse.ArgumentParser(description='Read SnpEff tsv file, extract values and write to new csv')

    parser.add_argument('--in', help='SnpEff file (tsv)', dest="input", required=True)
    parser.add_argument("--out", help="output file (csv)", dest="output", required=True)
    parser.add_argument("--version", action="version", version="0.0.1")

    return parser.parse_args()
    
def main():
    logging.config.dictConfig(LogConfig().stdout_config)
    logger = logging.getLogger("rinc.etl.process_snpeff")

    args = _parse_args()

    ps = ProcessSnpeff()
    
    variant_transcripts = ps.get_variant_transcripts(args.input)
        
    ps.write(args.output, variant_transcripts)

    logger.info(f"Wrote {len(variant_transcripts)} variant transcripts to {args.output}")


if __name__ == '__main__':
    main()